### PR TITLE
CDAP-20321: Add Push Operation for InMemoryGitOperationRunner

### PR DESCRIPTION
**Pull request by samdgupi**
- Added necessary interfaces and classes for operation runner
- Added base implementation for InMemoryGitOperationRunner
- Added push app implementation

TODO
- Make the operations synchronized for a namespace

Testing
- Unit

### CODE REVIEW

File: cdap-app-fabric/pom.xml [hunk 1]
```diff
@@ -264,6 +264,12 @@
       <groupId>org.slf4j</groupId>
       <artifactId>log4j-over-slf4j</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-source-control</artifactId>
+      <version>6.9.0-SNAPSHOT</version>
```
Comment (de-lan) [inline cdap-app-fabric/pom.xml]: nit: use `${project.version}`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/AppDetailsToPush.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  public String getApplicationSpecString() {
+    return applicationSpecString;
+  }
+
+  public String getApplicationName() {
+    return applicationName;
+  }
+}
```
Comment (de-lan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/AppDetailsToPush.java:41]: end of line should be a new line

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/SourceControlManagementManagementServiceTest.java [hunk 3]
```diff
@@ -33,7 +33,7 @@
 /**
  *
  */
-public class SourceControlManagementServiceTest extends AppFabricTestBase {
+public class SourceControlManagementManagementServiceTest extends AppFabricTestBase {
```
Comment (de-lan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/SourceControlManagementManagementServiceTest.java]: Two `Management`s?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/SourceControlManagementManagementHttpHandlerTests.java [hunk 4]
```diff
@@ -35,7 +35,7 @@
 /**
  * Tests for {@link SourceControlManagementHttpHandler}
  */
-public class SourceControlManagementHttpHandlerTests extends AppFabricTestBase {
+public class SourceControlManagementManagementHttpHandlerTests extends AppFabricTestBase {
```
Comment (de-lan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/SourceControlManagementManagementHttpHandlerTests.java]: Two `Management`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlOperationFactoryProvider.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    this.cConf = cConf;
+    this.injector = injector;
+  }
+
+  @Override
+  public SourceControlOperationRunnerFactory get() {
+    boolean workerPoolEnabled = cConf.getBoolean(Constants.TaskWorker.POOL_ENABLE);
+    if (workerPoolEnabled) {
+      return injector.getInstance(Key.get(SourceControlOperationRunnerFactory.class,
+                                          Names.named(Constants.AppFabric.FACTORY_IMPLEMENTATION_REMOTE)));
```
Comment (de-lan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlOperationFactoryProvider.java:45]: Let's add the remote implementation in a separate PR, now just return `Key.get(SourceControlOperationRunnerFactory.class,
                                        Names.named(Constants.AppFabric.FACTORY_IMPLEMENTATION_LOCAL))`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
```
Comment (de-lan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java:33]: Let's install the module in https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/AppFabricServiceRuntimeModule.java:

```
install(
        new FactoryModuleBuilder()
          .implement(SourceControlOperationRunner.class, InMemorySourceControlOperationRunner.class)
          .build(Key.get(SourceControlOperationRunnerFactory.class,
                         Names.named(AppFabric.FACTORY_IMPLEMENTATION_LOCAL)))
      );
bind(SourceControlOperationRunnerFactory.class).toProvider(SourceControlOperatorFactoryProvider.class);
```

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/ListAppResponse.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  public String getApplicationName() {
+    return applicationName;
+  }
+
+  public String getApplicationFileHash() {
+    return applicationFileHash;
+  }
+}
```
Comment (de-lan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/ListAppResponse.java]: should be new line in end of line

File: pom.xml [hunk 8]
```diff
@@ -188,7 +188,9 @@
     <sonar.organization>cdap</sonar.organization>
     <sonar.projectKey>cdapio_${project.artifactId}</sonar.projectKey>
     <sonar.host.url>https://sonarcloud.io</sonar.host.url>
-    <sonar.coverage.jacoco.xmlReportPaths>${maven.multiModuleProjectDirectory}/cdap-coverage/target/site/jacoco-aggregate/jacoco.xml</sonar.coverage.jacoco.xmlReportPaths>
+    <sonar.coverage.jacoco.xmlReportPaths>
```
Comment (de-lan) [inline pom.xml]: Are these changes on purpose? Pls revert if not

File: cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
+///*
```
Comment (de-lan) [inline cdap-source-control/src/test/java/io/cdap/cdap/sourcecontrol/LocalGitServer.java]: Why comment out the whole file?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  private final RepositoryManager repositoryManager;
+  private final CConfiguration cConf;
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(CConfiguration cConf, RepositoryManager repositoryManager) {
```
Comment (de-lan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java:40]: So is this `RepositoryManager` `@Assisted` injected?

Comment (arjan-bal) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java:40]: RespositoryManager will be provided by guice?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    this.cConf = cConf;
+    this.repositoryManager = repositoryManager;
+  }
+
+  @Override
+  public ListenableFuture<PushAppsResponse> push(List<AppDetailsToPush> appsToPush, CommitMeta commitDetails,
+                                                 @Nullable String branchName) throws PushFailureException {
+
```
Comment (arjan-bal) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java:45]: nit: Blank line at the beginning of method.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    Path repoBasePath = repositoryManager.getBasePath();
+
+    //TODO: handle if appsToPush is too large
+    LOG.info("Pushing application configs for : {}", appsToPush.stream().map(AppDetailsToPush::getApplicationName));
+
+    Map<String, Path> applicationPathMap =
+      appsToPush.stream().collect(Collectors.toMap(AppDetailsToPush::getApplicationName,
+                                                   appDetail -> repoBasePath.resolve(appDetail.getApplicationName())));
```
Comment (arjan-bal) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java]: Do we want to create a directory for each artifact type (eg: streaming, data pipeline)?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java (hunk 10 above)
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java:40]: I think it's injected at runtime using `@Assist`, we're writing it this way because we also need to implement `RemoteSourceControlOperationRunner`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    this.repositoryManager = repositoryManager;
+  }
+
+  @Override
+  public ListenableFuture<PushAppsResponse> push(List<AppDetailsToPush> appsToPush, CommitMeta commitDetails,
+                                                 @Nullable String branchName) throws PushFailureException {
+
+    try {
+      repositoryManager.startAndWait();
+      repositoryManager.switchToCleanBranch(branchName);
```
Comment (arjan-bal) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java]: We should wrap the (switchToClean branch + file write + push call) in a function that is called with exponential backoff if there's a merge conflict.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java (hunk 12 above)
Comment (arjan-bal) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java]: Can you also make a method for getting the repo path based on appName? It will be used by pull also.
```
 private Path getAppPathInReposiroty(String appName) {
    return repositoryManager.getBasePath().resolve(appName);
  }
```

**Comment by samdgupi** [comment]
Changes merged in https://github.com/cdapio/cdap/pull/14911
